presto-main-base/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeJoinForSortedInputOptimizer.java

```diff
@@ -102,62 +109,63 @@ public boolean isPlanChanged()
         @Override
         public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
         {
-            // As of now, we only support inner join for merge join
-            if (node.getType() != INNER) {
-                return node;
+            PlanNode rewrittenLeft = node.getLeft().accept(this, context);
+            PlanNode rewrittenRight = node.getRight().accept(this, context);
+
+            boolean leftInputSorted = isPlanOutputSortedByColumns(rewrittenLeft, node.getCriteria().stream().map(EquiJoinClause::getLeft).collect(toImmutableList()));
+            boolean rightInputSorted = isPlanOutputSortedByColumns(rewrittenRight, node.getCriteria().stream().map(EquiJoinClause::getRight).collect(toImmutableList()));
+
+            if ((!leftInputSorted && !rightInputSorted) || (!prestoOnSpark && (!leftInputSorted || !rightInputSorted))) {
```

Comment:
1. Do not do merge join is no input is sorted
2. If only one side is sorted, do not do merge join if not presto on spark